research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+  with utils.logged_timer('%d games' % params.eval_games):
+    winner = evaluation.play_match(
+        black_net, white_net, params.eval_games, params.eval_readouts,
+        evaluate_dir, params.eval_verbose)
+
+  return black_model_name if winner == 'B' else white_model_name
+
+
+def main(_):
+  """Run the reinforcement learning loop."""
+  tf.logging.set_verbosity(tf.logging.INFO)
+
+  params = model_utils.MiniGoModelParams()
+
+  k = utils.round_power_of_two(FLAGS.go_size ** 2 / 3)
+  params.num_filters = k  # Number of filters in the convolution layer
+  params.fc_width = 2 * k  # Width of each fully connected layer
+  params.num_shared_layers = FLAGS.go_size  # Number of shared trunk layers
+  params.go_size = FLAGS.go_size  # Board size
+
+  # set the shuffle buffer size smaller
+  preprocessing.SHUFFLE_BUFFER_SIZE = 200000
+
+  # How many positions can fit on a graphics card. 256 for 9s, 16 or 32 for 19s.
+  if FLAGS.go_size == 9:
+    params.batch_size = 256
+  else:
+    params.batch_size = 32
+
+  # A dummy model for debug/testing purpose with fewer games and iterations
+  if FLAGS.debug:
+    params.num_filters = 8  # Number of filters in the convolution layer
+    params.fc_width = 16  # Width of each fully connected layer
+    params.num_shared_layers = 1  # Number of shared trunk layers
+    params.batch_size = 16
+    params.examples_per_generation = 64
+    params.max_games_per_generation = 10
+    params.max_iters_per_pipeline = 2
+    preprocessing.SHUFFLE_BUFFER_SIZE = 1000
+
+  # Set directories for models and datasets
+  if os.path.isdir(FLAGS.base_dir):  # if it already exists, delete it.
+    shutil.rmtree(FLAGS.base_dir, ignore_errors=True)
+  models_dir = os.path.join(FLAGS.base_dir, 'models')
+  estimator_working_dir = os.path.join(FLAGS.base_dir, 'estimator_working_dir/')
+  selfplay_dir = os.path.join(FLAGS.base_dir, 'data/selfplay/')
+  holdout_dir = os.path.join(FLAGS.base_dir, 'data/holdout/')
+  training_chunk_dir = os.path.join(FLAGS.base_dir, 'data/training_chunks/')
+  sgf_dir = os.path.join(FLAGS.base_dir, 'sgf/')
+  evaluate_dir = os.path.join(FLAGS.base_dir, 'sgf/evaluate/')
+
+  print('Creating random initial weights...')
+  bootstrap(estimator_working_dir, models_dir, params)
+
+  # Start from the bootstrap model to play games.
+  print('Playing some games...')
+  model_name = '000000-bootstrap'
+  for rl_iter in range(params.max_iters_per_pipeline):
+    print('RL_iteration: %d' % rl_iter)
+
+    # Self-play to generate at least params.max_games_per_generation games
+    while True:
+      params.holdout_pct = 0.05
+      selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
+               params)
+      if FLAGS.holdout:
+        params.holdout_pct = 1
+        selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
+                 params)
+      games = tf.gfile.Glob(os.path.join(selfplay_dir, model_name, '*.zz'))
+      if len(games) >= params.max_games_per_generation:
```

Comment:
I fear while True loops. Why is this necessary? When is the number of games less than one? When is the number of games > 1? Would this work if we did `for _ in params.max_games_per_generation` and then break early if the num games grows faster than one per generation?